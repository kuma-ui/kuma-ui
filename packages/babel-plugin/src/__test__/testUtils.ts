import { BabelFileResult, transformSync } from "@babel/core";
import plugin from "../";

export function babelTransform(
  code: string,
  runtime?: "classic" | "automatic"
): BabelFileResult {
  const filename = "test.tsx";
  const result1 = transformSync(code, {
    plugins: [plugin],
    filename,
  });

  if (result1 === null || result1.code == null)
    throw new Error(`Could not transform`);

  if (!runtime) {
    return result1;
  }

  const result2 = transformSync(result1.code, {
    presets: [
      "@babel/preset-typescript",
      [
        "@babel/preset-react",
        {
          runtime,
        },
      ],
    ],
    filename,
  });

  if (result2 === null || result2.code == null)
    throw new Error(`Could not transform`);

  return {
    metadata: result1.metadata,
    code: result2.code,
  };
}

export function getExpectSnapshot(result: BabelFileResult) {
  return `
${(result.metadata as { css: string } | undefined)?.css}

${result.code}
`;
}
