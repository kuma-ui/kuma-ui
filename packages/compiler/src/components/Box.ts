import type { NodePath, PluginPass, PluginObj, types as t } from "@babel/core";

export const handleBox = (path: NodePath<t.JSXElement>) => {};
