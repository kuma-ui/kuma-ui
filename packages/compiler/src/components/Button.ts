import { NodePath, PluginPass, PluginObj, types as t } from "@babel/core";

export const handleButton = (path: NodePath<t.JSXElement>) => {
  const openingElement = path.node.openingElement;
  if (!t.isJSXIdentifier(openingElement.name)) return;

  // add 'cursor="pointer"' attribute if it's not already present
  if (
    !openingElement.attributes.some(
      (attr) => attr.type === "JSXAttribute" && attr.name.name === "cursor"
    )
  ) {
    openingElement.attributes.push(
      t.jsxAttribute(t.jsxIdentifier("cursor"), t.stringLiteral("pointer"))
    );
  }
};
