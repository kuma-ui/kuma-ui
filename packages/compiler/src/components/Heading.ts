import type { NodePath, PluginPass, PluginObj, types as t } from "@babel/core";

export const handleHeading = (path: NodePath<t.JSXElement>) => {};
