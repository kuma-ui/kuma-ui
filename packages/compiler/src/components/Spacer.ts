import { NodePath, types as t } from "@babel/core";

/**
 * Transforms a Spacer JSXElement to include specific style attributes.
 * Removes 'size' and 'horizontal' attributes and adds 'display', 'flexShrink',
 * 'width', and 'height' attributes based on the original attributes.
 *
 * @param {NodePath<t.JSXElement>} path - The JSXElement NodePath to transform.
 */
export const handleSpacer = (path: NodePath<t.JSXElement>) => {
  const openingElement = path.node.openingElement;
  if (!t.isJSXIdentifier(openingElement.name, { name: "Spacer" })) return;

  // get 'size' and 'horizontal' props
  let size: t.JSXAttribute | undefined;
  let horizontal: t.JSXAttribute | undefined;

  openingElement.attributes = openingElement.attributes.filter((attr) => {
    if (t.isJSXAttribute(attr)) {
      if (attr.name.name === "size") {
        size = attr;
        return false;
      } else if (attr.name.name === "horizontal") {
        horizontal = attr;
        return false;
      }
    }
    return true;
  });

  // define attributes for vertical or horizontal spacer
  const displayValue = horizontal ? "inline-block" : "block";
  const display = openingElement.attributes.some(
    (attr) => t.isJSXAttribute(attr) && attr.name.name === "display"
  )
    ? undefined
    : t.jsxAttribute(t.jsxIdentifier("display"), t.stringLiteral(displayValue));

  const flexShrink = openingElement.attributes.some(
    (attr) => t.isJSXAttribute(attr) && attr.name.name === "flexShrink"
  )
    ? undefined
    : t.jsxAttribute(
        t.jsxIdentifier("flexShrink"),
        t.jSXExpressionContainer(t.numericLiteral(0))
      );

  let widthValue, heightValue;
  if (size && t.isJSXExpressionContainer(size.value)) {
    widthValue = horizontal
      ? size.value
      : t.jSXExpressionContainer(t.stringLiteral("auto"));
    heightValue = horizontal
      ? t.jSXExpressionContainer(t.stringLiteral("auto"))
      : size.value;
  } else {
    widthValue = horizontal
      ? t.jSXExpressionContainer(t.numericLiteral(0))
      : t.jSXExpressionContainer(t.stringLiteral("auto"));
    heightValue = horizontal
      ? t.jSXExpressionContainer(t.stringLiteral("auto"))
      : t.jSXExpressionContainer(t.numericLiteral(0));
  }

  const width = openingElement.attributes.some(
    (attr) => t.isJSXAttribute(attr) && attr.name.name === "width"
  )
    ? undefined
    : t.jsxAttribute(t.jsxIdentifier("width"), widthValue);

  const height = openingElement.attributes.some(
    (attr) => t.isJSXAttribute(attr) && attr.name.name === "height"
  )
    ? undefined
    : t.jsxAttribute(t.jsxIdentifier("height"), heightValue);

  // add attributes
  [display, flexShrink, width, height].forEach((attr) => {
    if (attr) openingElement.attributes.push(attr);
  });
};
