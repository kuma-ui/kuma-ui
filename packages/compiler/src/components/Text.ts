import type { NodePath, PluginPass, PluginObj, types as t } from "@babel/core";

export const handleText = (path: NodePath<t.JSXElement>) => {};
