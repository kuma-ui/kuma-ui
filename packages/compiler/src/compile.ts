import {
  Project,
  Node,
  SyntaxKind,
  JsxOpeningElement,
  JsxSelfClosingElement,
} from "ts-morph";
import { collectPropsFromJsx } from "./collector";
import { extractProps } from "./extractor";
import { componentList } from "@kuma-ui/core/components/componentList";
import { optimize } from "./optimizer/optimize";
import { processTaggedTemplateExpression } from "./processTaggedTemplateExpression";

const project = new Project({});

const compile = (
  code: string,
  id: string,
  bindings: Record<string, string>
) => {
  const css: string[] = [];
  const source = project.createSourceFile(id, code, { overwrite: true });
  source.forEachDescendant((node) => {
    if (
      node.getKind() === SyntaxKind.JsxElement ||
      node.getKind() === SyntaxKind.JsxSelfClosingElement
    ) {
      let openingElement: JsxOpeningElement | JsxSelfClosingElement;
      if (node.getKind() === SyntaxKind.JsxElement) {
        const jsxElement = node.asKindOrThrow(SyntaxKind.JsxElement);
        openingElement = jsxElement.getOpeningElement();
      } else {
        openingElement = node.asKindOrThrow(SyntaxKind.JsxSelfClosingElement);
      }
      const jsxTagName = openingElement.getTagNameNode().getText();
      // Check if the current JSX element is a Kuma component
      const originalComponentName = Object.keys(bindings).find(
        (key) =>
          bindings[key] === jsxTagName &&
          Object.values(componentList).some((c) => c === key)
      );
      if (!originalComponentName) return;

      const componentName =
        originalComponentName as (typeof componentList)[keyof typeof componentList];
      const extractedPropsMap = collectPropsFromJsx(openingElement);
      const result = extractProps(
        componentName,
        openingElement,
        extractedPropsMap
      );
      if (result) css.push(result.css);

      optimize(componentName, openingElement, extractedPropsMap["as"]);
    }
    if (Node.isTaggedTemplateExpression(node)) {
      processTaggedTemplateExpression(node, bindings);
    }
  });
  return { code: source.getFullText(), id, css: css.join(" ") };
};

export { compile };
