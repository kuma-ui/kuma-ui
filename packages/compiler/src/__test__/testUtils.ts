import { compile } from "../compile";

type Result = ReturnType<typeof compile>;

export function getExpectSnapshot(result: Result) {
  return `
  ${result.css}
  
  ${result.code}
  `;
}
