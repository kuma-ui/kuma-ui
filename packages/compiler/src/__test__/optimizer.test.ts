import { compile } from "../compile";
import { getExpectSnapshot } from "./testUtils";
import { describe, it, expect } from "vitest";
import { componentList } from "@kuma-ui/core/components/componentList";

describe("optimizer", () => {
  it("should compile Kuma components to raw HTML when no dynamic props are given", () => {
    // Arrange
    const code = `
        import { Box, Text } from "@kuma-ui/core";
        
        function App() {
          const red = "red";
          return (
            <HStack flexDir={["row", "column"]}>
              <Text>hello</Text>
            </HStack>
          );
        }
        `;

    // Act
    const result = compile(code, "file.tsx", componentList);
    // Assert
    expect(getExpectSnapshot(result)).toMatchSnapshot();
  });

  it("should not compile Kuma components with dynamic props to raw HTML", () => {
    // Arrange
    const code = `
      import { Box, Text, HStack } from "@kuma-ui/core";
      import { useState } from 'react'
      
      function App() {
        const red = "red"
        return (
          <HStack color={true ? red : 'blue'}>
            <Text>hello</Text>
          </HStack>
        );
      }
      `;

    // Act
    const result = compile(code, "file.tsx", componentList);

    // Assert
    expect(getExpectSnapshot(result)).not.toMatchSnapshot();
  });

  it("should not compile to raw HTML when spread attributes are present", () => {
    // Arrange
    const code = `
      import { Box, Text, HStack } from "@kuma-ui/core";
      
      function App(props) {
        return (
          <HStack {...props}>
            <Text>hello</Text>
          </HStack>
        );
      }
      `;

    // Act
    const result = compile(code, "file.tsx", componentList);

    // Assert
    expect(getExpectSnapshot(result)).not.toMatchSnapshot();
  });

  it("should compile SelfClosingTags with static styles to raw HTML", () => {
    // Arrange
    const code = `
      import { Image } from "@kuma-ui/core";
      
      function App() {
        return (
          <Image src="path/to/image.jpg" alt="Sample Image" />
        );
      }
      `;

    // Act
    const result = compile(code, "file.tsx", componentList);

    // Assert
    expect(getExpectSnapshot(result)).toMatchSnapshot();
  });
});
