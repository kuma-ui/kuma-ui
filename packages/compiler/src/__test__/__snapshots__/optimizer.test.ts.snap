// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`optimizer > should compile Kuma components to raw HTML when no dynamic props are given 1`] = `
"
  .üêª-4011430064 { display: flex;flex-direction: row; }@media (min-width: 576px) { .üêª-4011430064 { flex-direction: column; } } 
  
  
        import { Box, Text } from \\"@kuma-ui/core\\";
        
        function App() {
          const red = \\"red\\";
          return (
            <div className={\\"üêª-4011430064\\"}>
              <p>hello</p>
            </div>
          );
        }
        
  "
`;

exports[`optimizer > should compile SelfClosingTags with static styles to raw HTML 1`] = `
"
  
  
  
      import { Image } from \\"@kuma-ui/core\\";
      
      function App() {
        return (
          <img src=\\"path/to/image.jpg\\" alt=\\"Sample Image\\" />
        );
      }
      
  "
`;

exports[`optimizer > should not compile Kuma components with dynamic props to raw HTML 1`] = `
"
  .üêª-2037632742 { display: flex;flex-direction: row; } 
  
  
      import { Box, Text, HStack } from \\"@kuma-ui/core\\";
      import { useState } from 'react'
      
      function App() {
        const red = \\"red\\"
        return (
          <HStack color={true ? red : 'blue'} className={\\"üêª-2037632742\\"}>
            <p>hello</p>
          </HStack>
        );
      }
      
  "
`;

exports[`optimizer > should not compile to raw HTML when spread attributes are present 1`] = `
"
  .üêª-2037632742 { display: flex;flex-direction: row; } 
  
  
      import { Box, Text, HStack } from \\"@kuma-ui/core\\";
      
      function App(props) {
        return (
          <HStack {...props} className={\\"üêª-2037632742\\"}>
            <p>hello</p>
          </HStack>
        );
      }
      
  "
`;
